### ISSUE TITLE
Issue with TabMixPlus extension.

### ISSUE THREAD

**Issue by the-lamer** [Feodor2/Mypal#92]
Hello,
I've just migrated from Firefox 52.9.0esr to Mypal 28.7.1 on WindowsXP and found my most important extention TabMixPlus is working on Mypal, but with some issues (looks related to browser itself).
Extension behaviour:
TabMixPlus is an extension to make multiple rows on tabs panel. It has configuration of tab title width limits (min and max) and acts on next way: when tab row is full, it starts to reduce tab titles width to free space for new tab. If "min" tab width limit is reached, it adds new tabs row below curent one and place new tab on it. When second row will be full of tabs - it will add third row and so on. 
When user closes some tabs and one additional row becomes empty - tabmix removes this row.
Problem description:
1) Tabmixplus fails to add new row more than once after browser window opening (browser startup or new window creation).
2) Tabmixplus fails to remove empty rows on some tabs closure. 
Overall it looks like Tabmix attempt to modify rows quantity ends with hidden exception on first time and then new modifying request can't be processed at all.
Steps to reproduce:
1) Install Mypal 28.7.1 on WindowsXP SP3 32bit, install TabMixPlus  0.5.8.1 (latest available on built-in extention search in Mypal).
2) Open new browser window and clone current tab multiple times until new tabs row will appear.
3) Continue cloning tabs until new row will be full. 
Issue1: new tabs row not appearing. New tabs can be created, but not visible on tabs rows (you can see them when you close some early created tabs). 
4) Close all tabs except one.
Issue2: two remaining tabs rows will be still displayed. Second one will be empty.
5) Repeat step3 until new tabs will stop to appear. Clone several more tabs (this will be not visible).
6) Close Mypal and start it once again, allowing TabMixPlus to restore last session.
Issue3: You will see enough rows (3 or even more) to display all tabs you created on step5.
7) Close all tabs except one.
Issue4 (similar to Issue2): you will see all (3 or even more) rows still displayed, but empty. 

Attached screenshots:
Step3 illustration. No more rows appearing when 2nd one is full.
![tabmix_2rows](https://user-images.githubusercontent.com/55839748/65690609-c30e5980-e06f-11e9-94a3-ef83213ca18d.PNG)
Step4 illustration: Second row is still displayed after tabs closure.
![tabmix_tabs_closed](https://user-images.githubusercontent.com/55839748/65690611-c3a6f000-e06f-11e9-8142-ceb01cb2b1a4.PNG)
Step6 illustration: Three tab rows created on session restore on startup.
![tabmix_tabs_restored_on_restart](https://user-images.githubusercontent.com/55839748/65690612-c3a6f000-e06f-11e9-87f0-dc7dce174310.PNG)
TabMixPlus extention version illustration.
![tabmix_ver](https://user-images.githubusercontent.com/55839748/65690614-c3a6f000-e06f-11e9-84e0-559bc255777d.PNG)

This extention works perfectly on Firefox 52.9.0esr, so it looks like the issue is in browsers code difference. If this can't be fixed in Mypal itself, i will try to contact TabmixPlus author, but i'm afraid this option will not help as, as far as i know, this extention had been developed for old Firefox only and now is discontinued.

**Comment by Feodor2** [comment]
>Open new browser window and clone current tab multiple times until new tabs row will appear.

How many times to clone? i did 50 and row not appeared.

**Comment by the-lamer** [comment]
> How many times to clone? i did 50 and row not appeared.
Depends of TabMixPlus minimum tab width settings and you screen width. (see attached screenshots to find this option). 
TabMixPlus behaviour:
- If you have only one tab opened in browser window, it will have width equal to "to" field value in settings.
- If you add one more tab - it will have the same width.
- If you continue adding tabs and next one can't fit tab row, all tabs width will be reduced to free space for new tab (exactly like default Firefox behaviour).
- If you continue adding tabs and their width reach the value specified in first field in settings (i.e. tab width reached user defined minimum) - TabMixPlus should add new tabs row and place new tab on it. 
- When you have browser window with a lot of tabs and thus a lot of rows and start closing tabs one by one, TabMixPlus will remove additional rows as soon as they become empty.

> By the way must to use clone item only or not?
No, any action which opens new tab will be ok. Cloning is just the simplest way for testing.

> 50 clones - no new tabs row appeared. 
If you have no ultra-wide screen and ultra-low value in TabMix tab width settings - you reproduced the issue.

**Comment by the-lamer** [comment]
One more notice, which might help with issue investigation.
Steps:
1. I opened new browser window with 1 search engine tab in it. 
2. I opened empty new tab in this window next to first one.
3. I cloned first tab with search endine 8 times (9 tabs total) and due to my screen width and min tab width setting in TabMix i got first row full and second row appeared.
4. I continued cloning until second row got full and last empty tab disappeared, because it have to be moved onto third row in this case, but third row was not visible.
![tabmix_3rows_1](https://user-images.githubusercontent.com/55839748/66127795-cdd76a00-e5ec-11e9-9762-61e6c0fbdd01.PNG)
5. I placed my mouse pointer on tabs bar and rotated scroll wheel down. And magiс happened: first tabs row disappeared in the top, but third tab row appeared in the bottom with tabs moved onto it. I cloned "empty new tab" several times to ensure this row is fully functional. 
![tabmix_3rows_2](https://user-images.githubusercontent.com/55839748/66127796-cdd76a00-e5ec-11e9-84b2-b822d714608f.PNG)

So, the issue description now looks like: TabMixPlus extension forks fine in MyPal, but browser allows it to modify size of tab panel only once for one window. 
Maybe, some kind of tab panel complete re-draw may be done on each tab adding or removal event on browser layer, which will finally fix the issue?

**Comment by Feodor2** [comment]
Thanks for new data, i investigate an error in the error console if it is related.

**Comment by Feodor2** [comment]
[Tabmmm.zip](https://github.com/Feodor2/Mypal/files/3959439/Tabmmm.zip)

Try this modified version, had to disable some features:
Tabmix.setItem(tabsToolbar, "classic40", version);
gTMPprefObserver.updateTabsStyle("unloadedTab");
gTMPprefObserver.setProgressMeter();

**Comment by the-lamer** [comment]
Feodor, thank a lot, now it works fine!
Sorry for late reply, somehow i missed this thread notification.

Is it possible to at least partially restore "unloadedTab" highlight? Maybe, not a complete custom styling, but just color or font or even append of any specific character to such tabs title?

I'm widely using this TabMixPlus functionality - keeping tabs unloaded by default on session restore. This helps to survive on PC with relatively small amount of RAM while browsing heavy sites and keeping a lot of tabs opened in background temporary. Without this markup it's become hard to decide have i restart my browser prior to opening new tab with heavy site or not.

For example, i'm writing this post from PC with Pentium4 dual core CPU and 2GB RAM. Due to other applications i had only 1GB ram free prior to starting MyPal. I have a session with 24 tabs opened (single window), but only 6 of them are loaded (other kept non-loaded, background for future work), and MyPal consumes ~800MB RAM (so i have about 200MB headroom). In this case i will be able to open 10-15 new active tabs more with ultra-light sites (forums, plain html sites like wikipedia), but only 1-2 new active tabs with heavy sites (youtube, facebook or main page of news sites with a lot of videos and images). 
In the same situation if i have session with 24 tabs opened, but 20 of them loaded, i will be more likely already out of free ram and if I open new tab with heavy site it will lead to crash, so i have to restart MyPal prior to doing this action.
So it's really important to be able to see amount of loaded tabs each time.

**Comment by Feodor2** [comment]
i shall try
